tests/test_qnode.py

```diff
@@ -1838,6 +1851,19 @@ def circuit():
         results = circuit(shots=20)  # pylint: disable=unexpected-keyword-arg
         assert qml.math.allclose(results, np.zeros((20, 2)))
 
+    def test_wires_property(self):
+        """Test the wires property behaves as expected with the new device API"""
+        dev = experimental.DefaultQubit2(wires=3)
+
+        @qml.qnode(dev)
+        def circuit():
+            qml.Hadamard(0)
+            return qml.expval(qml.PauliZ(0))
+
+        assert circuit.wires is None  # with no tape, don't return any wires
+        circuit()  # construct a tape
+        assert circuit.wires is circuit.tape.wires
```

Comment:
I have mixed feelings about this behaviour. As you said yourself, the behaviour could be a bit unexpected for some users. I'm not sure if it's better or worse, but another option is to return device wires before tape construction. Whether or not that will be more confusing is up for debate

Reply:
I decided against it because I don't think it'll be confusing - I think it will be wrong. Consider this:
```python
@qml.qnode(qml.device("default.qubit.2", wires=100))
def circuit(wire):
    qml.Hadamard(wire)
    qml.CNOT([wire, 1])
    return qml.expval(qml.PauliZ(1))

do_something_with(circuit.wires)
```
it might be quite wasteful to do something with 100 wires when the real answer will eventually be `[wire, 1]` (which depends on QNode construction). I am open to raising an error and suggesting the user calls the QNode at least once before accessing `qnode.wires`

Reply:
Quick question :) What is the problem we are trying to solve for here?

Reply:
@josh146  We basically need to add a wire order to the device for `qml.state()`

For example:
```
@qml.qnode(qml.device('default.qubit'))
def circuit():
    qml.Hadamard(1)
    qml.S(0)
    return qml.state()
```

What should the state look like in this situation?  

Every other measurement process has a specified wire order, but state doesn't.

The naive solution is "give state a wire order".  But giving state a wire order turns it into a density matrix, which is something entirely different.

Reply:
I'm going to remove this from this PR, and we can talk about it again if we need to while I'm working on `qml.state()` improvements.